lib/internal/fs/dir.js

```diff
@@ -59,11 +61,19 @@ class Dir {
       throw new ERR_INVALID_CALLBACK(callback);
     }
 
+    if (this[kDirBufferedEntries].length > 0) {
+      const [ name, type ] = this[kDirBufferedEntries].splice(0, 2);
+      getDirent(this[kDirPath], name, type, callback);
```

Comment:
Should this be wrapped in a `setImmediate` for timing consistency?

Reply:
@Fishrock123 So … this is a good point, yes.

Most of the time this will be called from the async iterator or the promisified version, I’d expect, and in that case it doesn’t matter. For the cases where it *is* visible to the user, using `setImmediate()` would likely undo a lot of the performance improvements.

What I’ve just pushed uses `process.nextTick()` when there’s a chance of synchronous use. That means there isn’t always the exact same timing depending on whether there are buffered entries, but I think that’s okay.

Reply:
Using `setImmediate` does not seem to have any substantial performance difference compared to the current `nextTick` code?
```
                                                            confidence improvement accuracy (*)   (**)  (***)
 fs/bench-opendir.js mode='async' dir='lib' n=100                           1.58 %       ±2.17% ±2.90% ±3.82%
 fs/bench-opendir.js mode='async' dir='test/parallel' n=100                -0.05 %       ±0.56% ±0.74% ±0.96%
 fs/bench-opendir.js mode='sync' dir='lib' n=100                            1.15 %       ±3.46% ±4.60% ±5.98%
 fs/bench-opendir.js mode='sync' dir='test/parallel' n=100                  0.70 %       ±1.01% ±1.34% ±1.75%
```

Reply:
@Fishrock123 If you’re comparing with the current code in this PR – `maybeSync` is always false there, so the benchmarks don’t trigger that path.

Reply:
```diff
diff --git a/benchmark/fs/bench-opendir.js b/benchmark/fs/bench-opendir.js
index 84d8e6598774..cda1d3032151 100644
--- a/benchmark/fs/bench-opendir.js
+++ b/benchmark/fs/bench-opendir.js
@@ -7,7 +7,7 @@ const path = require('path');
 const bench = common.createBenchmark(main, {
   n: [100],
   dir: [ 'lib', 'test/parallel'],
-  mode: [ 'async', 'sync' ]
+  mode: [ 'async', 'sync', 'callback' ]
 });
 
 async function main({ n, dir, mode }) {
@@ -21,7 +21,23 @@ async function main({ n, dir, mode }) {
       // eslint-disable-next-line no-unused-vars
       for await (const entry of await fs.promises.opendir(fullPath))
         counter++;
-    } else {
+    } else if (mode === 'callback') {
+      const dir = await fs.promises.opendir(fullPath);
+      await new Promise((resolve, reject) => {
+        function read() {
+          dir.read((err, entry) => {
+            if (err)
+              reject(err);
+            if (entry === null)
+              resolve(dir.close());
+            else
+              read();
+          });
+        }
+
+        read();
+      });
+    } {
       const dir = fs.opendirSync(fullPath);
       while (dir.readSync() !== null)
         counter++;
diff --git a/lib/internal/fs/dir.js b/lib/internal/fs/dir.js
index fedd7ff8eddb..ab37391e14e7 100644
--- a/lib/internal/fs/dir.js
+++ b/lib/internal/fs/dir.js
@@ -70,7 +70,7 @@ class Dir {
     if (this[kDirBufferedEntries].length > 0) {
       const [ name, type ] = this[kDirBufferedEntries].splice(0, 2);
       if (maybeSync)
-        process.nextTick(getDirent, this[kDirPath], name, type, callback);
+        setImmediate(getDirent, this[kDirPath], name, type, callback);
       else
         getDirent(this[kDirPath], name, type, callback);
       return;
```

results in 

```
                                                                confidence improvement accuracy (*)    (**)   (***)
 fs/bench-opendir.js mode='callback' dir='lib' n=100                           2.10 %       ±9.26% ±12.33% ±16.05%
 fs/bench-opendir.js mode='callback' dir='test/parallel' n=100        ***    -29.21 %       ±5.86%  ±7.82% ±10.24%
```

for me